vm/qemu/qmp.go

```diff
@@ -53,29 +48,45 @@ func (inst *instance) qmpConnCheck() error {
 		return err
 	}
 
-	monDec := json.NewDecoder(conn)
+	scanner := bufio.NewScanner(conn)
 	monEnc := json.NewEncoder(conn)
 
-	var banner qmpBanner
-	if err := monDec.Decode(&banner); err != nil {
-		return err
-	}
+	inst.mon = conn
+	inst.scanner = scanner
+
+	inst.qmpRecv()
 
 	inst.monEnc = monEnc
-	inst.monDec = monDec
 	if _, err := inst.doQmp(&qmpCommand{Execute: "qmp_capabilities"}); err != nil {
 		inst.monEnc = nil
-		inst.monDec = nil
+		inst.mon = nil
+		inst.scanner = nil
 		return err
 	}
-	inst.mon = conn
 
 	return nil
 }
 
 func (inst *instance) qmpRecv() (*qmpResponse, error) {
 	qmp := new(qmpResponse)
-	err := inst.monDec.Decode(qmp)
+	var err error
+
+	for inst.scanner.Scan() {
```

Comment:
It took me a bit to grasp the logic here, I subtly mis-read it several times.
Is this equivalent code?

```
func (inst *instance) qmpRecv() (*qmpResponse, error) {
	var err error
	for inst.scanner.Scan() {
		b := inst.scanner.Bytes()
		qe := new(qmpEvent)
		err = json.Unmarshal(b, &qe)
		if err != nil {
			continue
		}
		if qe.Event != "" {
			continue  // events are not interesting for us
		}
		qmp := new(qmpResponse)
		err = json.Unmarshal(b, qmp)
		if err != nil {
			continue
		}
		return qmp, nil
	}
	return nil, err
}
```

I think it's a bit better as it makes qmp variable more local, reduces nesting and does not return non-nil qmp and error at the same time (which is I think the real root cause of the crash fixed in the other commit).

I think this may have an issue if the very first `inst.scanner.Scan()` returns false (which is quite expected, if the connection fails in between messages), then the error is in s.Err().
Can it also discard the real error if there is an empty line in between messages?
Can it also discard the real error if we failed to parse a response, but then there is an event follows it?

I guess I still can't fully understand how it will behave in all cases...

It may be simple if we keep json.Decoder (don't add Scanner) and extend qmpResponse with Event:

```
type qmpResponse struct {
	Event     string
	Error struct {
		Class string
		Desc  string
	}
	Return interface{}
}
```

Then: decode a response, if it has Event, then decode another one.